examples/audio/ctc_asr.py

```diff
@@ -0,0 +1,514 @@
+    # 2. Decode the wav file
+    audio, _ = tf.audio.decode_wav(file)
+    audio = tf.squeeze(audio, axis=-1)
+    # 3. Compute the padding tensor
+    zero_padding = tf.zeros([maxlen_x] - tf.shape(audio), dtype=tf.float32)
+    # 4. Change type to float
+    audio = tf.cast(audio, tf.float32)
+    # 5. Pad the audio
+    equal_length = tf.concat([audio, zero_padding], 0)
+    # 6. Get the spectrogram
+    spectrogram = tf.signal.stft(
+        equal_length, frame_length=frame_length, frame_step=frame_step
+    )
+    # 7. We only need the magnitude, which can be derived by applying tf.abs
+    spectrogram = tf.abs(spectrogram)
+    # 8. Add dimension
+    # 9. Map the characters in label to numbers
+    label = char_to_num(tf.strings.unicode_split(label, input_encoding="UTF-8"))
+    # 10. Return a dict as our model is expecting two inputs
+    return spectrogram, label
+
+
+"""
+We now create our `tf.data.Dataset` object returns a new dataset
+containing the transformed elements, in the same order as they
+appeared in the input. The function `encode_single_sample` is
+used to change both the values and the structure of a dataset's elements.
+
+"""
+
+# Define the batch size
+batch_size = 128
+
+# Define the trainig dataset
+train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
+train_dataset = (
+    train_dataset.map(encode_single_sample, num_parallel_calls=tf.data.AUTOTUNE)
+    .batch(batch_size)
+    .prefetch(buffer_size=tf.data.AUTOTUNE)
+)
+
+# Define the validation dataset
+validation_dataset = tf.data.Dataset.from_tensor_slices((x_valid, y_valid))
+validation_dataset = (
+    validation_dataset.map(encode_single_sample, num_parallel_calls=tf.data.AUTOTUNE)
+    .batch(batch_size)
+    .prefetch(buffer_size=tf.data.AUTOTUNE)
+)
+
+
+"""
+## Visualize the data
+
+Let's visualize an example in our dataset, including the
+audio clip, the spectrogram and the corresponding label.
+"""
+
+fig = plt.figure(figsize=(20, 5))
+for batch in train_dataset.shuffle(buffer_size=16).take(1):
+    spectrograms = batch[0]
+    labels = batch[1]
+    for i in range(1):
+        input_shape = (spectrograms[i].shape[0], spectrograms[i].shape[1])
+        label = tf.strings.reduce_join(num_to_char(labels[i])).numpy().decode("utf-8")
+        ax = plt.subplot(2, 1, 1)
+        ax.imshow(np.transpose(spectrograms[i]), vmax=1)
+        ax.set_title(label.replace("#", ""))
+        ax.axis("off")
+        # Convert to wave format
+        spectrograms_ = tf.expand_dims(spectrograms[i], axis=2)
+        audio = tf.signal.inverse_stft(
+            tf.cast(spectrograms[i], tf.complex128), frame_length, frame_step
+        )
+        ax = plt.subplot(2, 1, 2)
+        plt.plot(audio)
+        ax.set_title("Signal Wave")
+        ax.set_xlim(0, len(audio))
+        display.display(display.Audio(audio, rate=16000))
+plt.show()
+
+
+"""
+## Model
+"""
+
+"""
+We first define the CTC Loss function.
+"""
+
+
+def CTCLoss(y_true, y_pred):
```

Comment:
I'm thinking we should probably make a CTC loss part of the core Keras API. What do you think?

Reply:
I definitely agree with you! I will try to do it and I will submit a pull request to the main Keras library!